pallets/token-factory/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+				// from: H160
+				from,
+				// input: Vec<u8>
+				contract,
+				// value: U256
+				U256::zero(),
+				// gas limit: U256
+				U256::from(0x100000),
+				// gas price: U256
+				Some(U256::from(1)),
+				// nonce: Option<H256>
+				Some(U256::zero()),
+				// action: TransactionAction
+				pallet_ethereum::TransactionAction::Create,
+				// config: Option<EvmConfig>
+				None,
+			)? {
+				T::AccountId::decode(&mut &addr[..]).expect("AccountId: H160")
+			} else {
+				// Rust forces us to cover when no address is returned and execute
+				// call succeeds, but this never happens in practice with TransactionAction::Create
+				panic!("Unreachable branch; execute never succeeds without deployed address");
+			};
+			// update runtime storage
+			<ContractAddress<T>>::insert(id, address.clone());
+			<Tokens<T>>::mutate(|list| {
+				if let Err(loc) = list.binary_search(&id) {
+					list.insert(loc, id);
+				}
+			});
+			Self::deposit_event(Event::Registered(id, address));
+			Ok(().into())
+		}
+
+		#[pallet::weight(0)]
+		pub fn destroy_all(origin: OriginFor<T>, id: T::TokenId) -> DispatchResultWithPostInfo {
+			frame_system::ensure_root(origin)?;
+			let _address = <ContractAddress<T>>::get(id).ok_or(Error::<T>::IdNotClaimed)?;
+			// TODO: ethereum transaction to remove/kill contract
+			// clear storage and free id
+			<ContractAddress<T>>::remove(id);
+			<Tokens<T>>::mutate(|list| {
+				if let Ok(loc) = list.binary_search(&id) {
+					list.remove(loc);
+				}
+			});
+			// TODO: get this
+			let amount_destroyed = BalanceOf::<T>::zero();
+			Self::deposit_event(Event::DestroyedAll(id, amount_destroyed));
+			Ok(().into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn sudo_caller() -> H160 {
+			T::AccountToH160::convert(<pallet_sudo::Module<T>>::key())
+		}
+	}
+
+	pub trait TokenFactory<Id, Account, Balance> {
+		fn exists(id: Id) -> bool;
+		// setters
+		fn mint(id: Id, who: Account, amount: Balance);
+		fn burn(id: Id, who: Account, amount: Balance);
+		// getters
+		fn total_issuance(id: Id) -> Balance;
+		fn balance_of(id: Id, who: Account) -> Balance;
+	}
+
+	impl<T: Config> TokenFactory<T::TokenId, T::AccountId, BalanceOf<T>> for Pallet<T> {
+		fn exists(id: T::TokenId) -> bool {
+			<ContractAddress<T>>::get(id).is_some()
+		}
+		fn mint(id: T::TokenId, who: T::AccountId, amount: BalanceOf<T>) {
+			if let Some(address) = <ContractAddress<T>>::get(id) {
+				// TODO: mint by forming valid ethereum execute call using ABI
+				let mut input = hex_literal::hex!("9cff1ade").to_vec();
+				// append address
+				input.extend_from_slice(
+					H256::from(T::AccountToH160::convert(who.clone())).as_bytes(),
+				);
+				// append amount
+				input.extend_from_slice(
+					H256::from_uint(&U256::from(amount.saturated_into::<u128>())).as_bytes(),
+				);
+				// TODO: check if this means execution succeeded...I don't think it does lol
+				if let Ok(_) = <pallet_ethereum::Module<T>>::execute(
```

Comment:
What are the advantages of using this method rather than pallet_evm::call?

Reply:
None, good catch